moveit_experimental/jog_arm/src/jog_arm/jog_ros_interface.cpp

```diff
@@ -0,0 +1,1243 @@
+        if (incoming_jts_.position[m] != 0.)
+          all_zeros = false;
+        goto NEXT_JOINT;
+      }
+    }
+  NEXT_JOINT:;
+  }
+
+  return !all_zeros;
+}
+
+// Scale the incoming jog command
+Eigen::VectorXd JogCalcs::scaleCartesianCommand(const geometry_msgs::TwistStamped& command) const
+{
+  Eigen::VectorXd result(6);
+
+  // Apply user-defined scaling if inputs are unitless [-1:1]
+  if (parameters_.command_in_type == "unitless")
+  {
+    result[0] = parameters_.linear_scale * command.twist.linear.x;
+    result[1] = parameters_.linear_scale * command.twist.linear.y;
+    result[2] = parameters_.linear_scale * command.twist.linear.z;
+    result[3] = parameters_.rotational_scale * command.twist.angular.x;
+    result[4] = parameters_.rotational_scale * command.twist.angular.y;
+    result[5] = parameters_.rotational_scale * command.twist.angular.z;
+  }
+  // Otherwise, commands are in m/s and rad/s
+  else if (parameters_.command_in_type == "speed_units")
+  {
+    result[0] = command.twist.linear.x * parameters_.publish_period;
+    result[1] = command.twist.linear.y * parameters_.publish_period;
+    result[2] = command.twist.linear.z * parameters_.publish_period;
+    result[3] = command.twist.angular.x * parameters_.publish_period;
+    result[4] = command.twist.angular.y * parameters_.publish_period;
+    result[5] = command.twist.angular.z * parameters_.publish_period;
+  }
+  else
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Unexpected command_in_type");
+
+  return result;
+}
+
+Eigen::VectorXd JogCalcs::scaleJointCommand(const moveit_msgs::JogJoint& command) const
+{
+  Eigen::VectorXd result(jt_state_.name.size());
+
+  for (std::size_t i = 0; i < jt_state_.name.size(); ++i)
+  {
+    result[i] = 0.0;
+  }
+
+  // Store joints in a member variable
+  for (std::size_t m = 0; m < command.joint_names.size(); ++m)
+  {
+    for (std::size_t c = 0; c < jt_state_.name.size(); ++c)
+    {
+      if (command.joint_names[m] == jt_state_.name[c])
+      {
+        // Apply user-defined scaling if inputs are unitless [-1:1]
+        if (parameters_.command_in_type == "unitless")
+          result[c] = command.deltas[m] * parameters_.joint_scale;
+        // Otherwise, commands are in m/s and rad/s
+        else if (parameters_.command_in_type == "speed_units")
+          result[c] = command.deltas[m] * parameters_.publish_period;
+        else
+          ROS_ERROR_STREAM_NAMED(LOGNAME, "Unexpected command_in_type");
+        goto NEXT_JOINT;
+      }
+    }
+  NEXT_JOINT:;
+  }
+
+  return result;
+}
+
+// Calculate a pseudo-inverse.
+Eigen::MatrixXd JogCalcs::pseudoInverse(const Eigen::MatrixXd& J) const
+{
+  return J.transpose() * (J * J.transpose()).inverse();
+}
+
+// Add the deltas to each joint
+bool JogCalcs::addJointIncrements(sensor_msgs::JointState& output, const Eigen::VectorXd& increments) const
+{
+  for (std::size_t i = 0, size = static_cast<std::size_t>(increments.size()); i < size; ++i)
+  {
+    try
```

Comment:
Is it really necessary to check this all the time?
The size of those vectors shouldn't change over time, right?